modules/RTC/TPCUtils.js

```diff
@@ -671,122 +653,248 @@ export class TPCUtils {
-                    const transceiverMid = this.pc._localTrackTransceiverMids.get(oldTrack.rtcId);
-
-                    transceiver = this.pc.peerconnection.getTransceivers().find(t => t.mid === transceiverMid);
-                } else if (trackIndex) {
-                    transceiver = this.pc.peerconnection.getTransceivers()
-                            .filter(t => t.receiver.track.kind === mediaType
-                                && t.direction !== MediaDirection.RECVONLY)[trackIndex];
+    mungeCodecOrder(description) {
+        const codecSettings = this.pc.codecSettings;
+
+        if (!codecSettings) {
+            return description;
+        }
+
+        const { isP2P } = this.options;
+        const parsedSdp = transform.parse(description.sdp);
+        const mLines = parsedSdp.media.filter(m => m.type === codecSettings.mediaType);
+
+        if (!mLines.length) {
+            return description;
+        }
+
+        for (const mLine of mLines) {
+            const currentCodecs = this.getConfiguredVideoCodecs(description);
+
+            for (const codec of currentCodecs) {
+                if (isP2P) {
+                    // 1. Strip the high profile H264 codecs on all clients. macOS started offering encoder for H.264
+                    //   level 5.2 but a decoder only for level 3.1. Therfore, strip all main and high level codecs for
+                    //   H.264.
+                    // 2. There are multiple VP9 payload types generated by the browser, more payload types are added
+                    //   if the endpoint doesn't have a local video source. Therefore, strip all the high profile codec
+                    //   variants for VP9 so that only one payload type for VP9 is negotiated between the peers.
+                    if (codec === CodecMimeType.H264 || codec === CodecMimeType.VP9) {
+                        SDPUtil.stripCodec(mLine, codec, true /* high profile */);
+                    }
+
+                    // Do not negotiate ULPFEC and RED either.
+                    if (codec === CodecMimeType.ULPFEC || codec === CodecMimeType.RED) {
+                        SDPUtil.stripCodec(mLine, codec, false);
+                    }
+                }
+            }
+
+            // Reorder the codecs based on the preferred settings.
+            if (!this.pc.usesCodecSelectionAPI()) {
+                for (const codec of codecSettings.codecList.slice().reverse()) {
+                    SDPUtil.preferCodec(mLine, codec, isP2P);
                 }
             }
         }
-        if (!transceiver) {
-            return Promise.reject(
-                new Error(`Replace track failed - no transceiver for old: ${oldTrack}, new: ${newTrack}`));
-        }
-        logger.debug(`${this.pc} Replacing ${oldTrack} with ${newTrack}`);
 
-        return transceiver.sender.replaceTrack(track)
-            .then(() => Promise.resolve(transceiver));
+        return new RTCSessionDescription({
```

Comment:
The RTCSessionDescription constructor is deprecated and it's encouraged to use bare objects: https://developer.mozilla.org/en-US/docs/Web/API/RTCSessionDescription/RTCSessionDescription

Reply:
Do you know what version supports the bare objects? Since we need to check it across all browsers and add version checks, I suggest leaving it out of this one and doing it as a follow up PR. WDYT?

Reply:
All of them support bare object and have so for years AFAIK. Only the constructor was deprecated.